### ISSUE TITLE
discrete-time variability and access of public function components

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/ModelicaSpecification#83]
**Reported by PHarman on 21 Jan 2010 09:54 UTC**
Following is from Peter Fritzson email:

----

In Section 12.4.2       Output Formal Parameters of Functions
you can find the following example:

function eigen "calculate eigenvalues and optionally eigenvectors"
 parameter Integer n = size(A,1);
 input     Real    A[:, size(A,1)];
 output    Real    eigenValues[n,2];
 output    Real    rightEigenVectors[n,n];
 output    Real    leftEigenVectors [n,n];

We also have:

in Section 3.8.3

* Expressions in functions behave as though they were discrete-time
expressions.

and:

12.2    Function as a Specialized Class

* Each input formal parameter of the function must be prefixed by the
keyword input, and each result formal parameter by the keyword output. All
public variables are formal parameters.

What are the problems?

- The public parameter definition in eigen cannot be allowed since size(A,1)
has discrete-time variability according to 3.8.3
- There is also the problem with the statement  "All public variables are
formal parameters." in Section 12.2, since the Integer n is not a formal
parameter.

The proposed solution:

----

To allow the readable array dimension declarations in a function like
eigen, we should make the following changes in the specification:

In Section 12.4.2       Output Formal Parameters of Functions

function eigen "calculate eigenvalues and optionally eigenvectors"
 discrete Integer n = size(A,1);
 input     Real    A[:, n];
 output    Real    eigenValues[n,2];
 output    Real    rightEigenVectors[n,n];
 output    Real    leftEigenVectors [n,n];

Change in Section 12.2, the last statement in the bullet:

12.2    Function as a Specialized Class

* Each input formal parameter of the function must be prefixed by the
keyword input, and each result formal parameter by the keyword output. All
formal parameters are public.

It would also be useful to add the following:

in:  10.2       Flexible Array Sizes

An array dimension size can have at most parameter variability, except in
functions where it can have at most discrete-time variability.

New function restriction in Section 12.2:

* Public function components cannot be accessed by dot notation.

----

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/256

**Comment by modelica-trac-importer** [comment]
**Comment by PHarman on 21 Jan 2010 10:05 UTC**
Although we have the statement "Expressions in functions behave as though they were discrete-time expressions" we also have the statement "An array dimension size can have at most parameter variability", my understanding of which is that "size(A,1)" always has parameter variability.

I think prefixing "n" with discrete would be confusing to users because it is the intention that it is a parameter. We should clarify "Expressions in functions behave as though they were discrete-time expressions" because it is not clear what it applies to.

"Within a function, the following behave as though they have discrete-time variability:
- input and output variables
- components not declared as parameter or constant
- expressions involving the above components"

or something along those lines.

**Comment by modelica-trac-importer** [comment]
**Modified by PHarman on 21 Jan 2010 10:16 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 21 Jan 2010 10:26 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by PHarman on 25 Jan 2010 10:03 UTC**
Comment:
Harman:

"An array  dimension size can have at most parameter variability", my
understanding  of which is that "size(A,1)" always has parameter
variability.


PeterF:
This is wrong, and has been wrong for many years. The size of an
array in a function can be determined at run-time when the function is
called.

Peter F

**Comment by modelica-trac-importer** [comment]
**Comment by PHarman on 25 Jan 2010 10:15 UTC**
I'm not specifying whether the sizes are determined at runtime or compile time.

"The size of an array in a function can be determined" <- Yes, it can be determined, but during the function evaluation it doesn't change as we don't have a means of dynamically resizing arrays, either in functions or in models. So it's still not discrete (within the context of the function).

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 25 Jan 2010 11:49 UTC**
Actually array variables in functions can change size (see section 12.4.4 of Modelica 3.1 spec - supported interactively in Dymola).

Thus it seems that 'n' should not be a parameter (in MSL it is not present at all).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Mar 2010 11:49 UTC**
The "eigen" function example in Modelica 3.1 is wrong as stated in this ticket, and is corrected in the Modelica 3.2 draft specification

```
function eigen "calculate eigenvalues and optionally eigenvectors"
  input     Real    A[:, size(A,1)];
  output    Real    eigenValues[size(A,1),2];
  output    Real    rightEigenVectors[size(A,1),size(A,1)];
  output    Real    leftEigenVectors [size(A,1),size(A,1)];
```

It seems currently not high priority to introduce a mechanism to define an auxiliary variable as a dimension to make the further declarations easier to read. The "size(...)" is a bit cumbersome but works o.k.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 28 Apr 2010 07:12 UTC**
